### PR TITLE
THRIFT-3636 The precision is 15 bits after dot of casting from double to string in Thrift-cpp-library- json-protocol.

### PR DESCRIPTION
**Pull request by mygityf**
Cpp json protocol double precision should be 16 bits

### CODE REVIEW

**Comment by nsuke** [comment]
thanks for uncovering it !
The "2" is for "0." ?
I think the comment can be improved.
@hcorg maybe you have better alternative than code comment here ?

**Comment by mygityf** [comment]
The "1" is for "0.123456789012345".
The "2" is for "0.1234567890123456".

**Comment by nsuke** [comment]
ok, so 2 is the difference between max_digits10 and digits10.
since we cannot use c++11 max_digits10 yet, I suggest using 17 directly with a comment or a clearly named local variabe.

**Comment by mygityf** [comment]
using 17 directly with a comment.

**Comment by mygityf** [comment]
It looks no error detected for Jenkins checking.

**Comment by nsuke** [comment]
Actually test is failing quite badly.
Your fix is fine but the test needs fix too.
A possible fix is to replace the PI and golden ratio with float literals and use the same digits in the assertions.

File: lib/cpp/src/thrift/protocol/TJSONProtocol.cpp [hunk 1]
```diff
@@ -524,7 +524,8 @@ namespace {
 std::string doubleToString(double d) {
   std::ostringstream str;
   str.imbue(std::locale::classic());
-  str.precision(std::numeric_limits<double>::digits10 + 1);
+  // new precision is 17 equal to max_digits10(c++11) or digits10+2
+  str.precision(17);
```
Comment (hcorg) [inline lib/cpp/src/thrift/protocol/TJSONProtocol.cpp]: IMHO
`const double max_digits10 = 2 + std::numeric_limits<double>::digits10;`
`str.precision(max_digits10);`
looks better than comment ;)

I'd even consider naming variable `max_digits10_cpp11`

Comment (mygityf) [inline lib/cpp/src/thrift/protocol/TJSONProtocol.cpp]: It looks better than before.
Thanks.

**Comment by mygityf** [comment]
@nsuke I replace the PI and golden ratio with float literals and use the same digits(17) in file lib/cpp/test/JSONProtoTest.cpp.
The result looks OK in Jenkins checking.
why CI build failed.

**Comment by nsuke** [comment]
@mygityf the failure is unrelated to your patch so no need to worry about that any more.

Would you mind if I remove the comment as per @hcorg 's suggestion, now that we have nicely self explanatory variable ?

**Comment by mygityf** [comment]
@nsuke just do it to make it better.

**Pull request closed without merging**